cmd/nvidia-dra-plugin/vfio-device.go

```diff
@@ -0,0 +1,192 @@
+	return &VfioPciManager{
+		pciDevicesRoot:  "/sys/bus/pci/devices",
+		vfioDevicesRoot: "/dev/vfio",
+		sysModulesRoot:  "/sys/module",
+		driver:          vfioPciDriver,
+		vfioPciModule:   vfioPciModule,
+	}
+}
+
+// Init ensures the vfio-pci module is loaded on the host.
+func (vm *VfioPciManager) Init() error {
+	if !vm.isVfioPCIModuleLoaded() {
+		err := vm.loadVfioPciModule()
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (vm *VfioPciManager) isVfioPCIModuleLoaded() bool {
+	modules, err := os.ReadDir(vm.sysModulesRoot)
+	if err != nil {
+		return false
+	}
+
+	for _, module := range modules {
+		if module.Name() == vm.vfioPciModule {
+			return true
+		}
+	}
+
+	return false
+
+}
+
+func (vm *VfioPciManager) loadVfioPciModule() error {
+	cmd := exec.Command("modprobe", vm.vfioPciModule) //nolint:gosec
+	_, err := cmd.CombinedOutput()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Configure binds the GPU to the vfio-pci driver.
+func (vm *VfioPciManager) Configure(info *GpuInfo) error {
+	perGpuLock.Get(info.PciAddress).Lock()
+	defer perGpuLock.Get(info.PciAddress).Unlock()
+
+	driver, err := getDriver(vm.pciDevicesRoot, info.PciAddress)
+	if err != nil {
+		return err
+	}
+	if driver == vm.driver {
+		return nil
+	}
+	err = changeDriver(info.PciAddress, vm.driver)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Unconfigure binds the GPU to the nvidia driver.
+func (vm *VfioPciManager) Unconfigure(info *GpuInfo) error {
+	perGpuLock.Get(info.PciAddress).Lock()
+	defer perGpuLock.Get(info.PciAddress).Unlock()
+
+	driver, err := getDriver(vm.pciDevicesRoot, info.PciAddress)
+	if err != nil {
+		return err
+	}
+	if driver == nvidiaDriver {
+		return nil
+	}
+	err = changeDriver(info.PciAddress, nvidiaDriver)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func getDriver(pciDevicesRoot, pciAddress string) (string, error) {
+	driverPath, err := os.Readlink(filepath.Join(pciDevicesRoot, pciAddress, "driver"))
+	if err != nil {
+		return "", err
+	}
+	_, driver := filepath.Split(driverPath)
+	return driver, nil
+}
+
+func changeDriver(pciAddress, driver string) error {
+	err := unbindFromDriver(pciAddress, driver)
+	if err != nil {
+		return err
+	}
+	err = bindToDriver(pciAddress, driver)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func unbindFromDriver(pciAddress, driverResetRetries string) error {
+	cmd := exec.Command(unbindFromDriverScript, pciAddress, driverResetRetries) //nolint:gosec
+	_, err := cmd.CombinedOutput()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func bindToDriver(pciAddress, driver string) error {
+	cmd := exec.Command(bindToDriverScript, pciAddress, driver) //nolint:gosec
+	_, err := cmd.CombinedOutput()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (vm *VfioPciManager) getIommuGroupForVfioPciDevice(pciAddress string) string {
+	iommuGroup, err := os.Readlink(filepath.Join(vm.pciDevicesRoot, pciAddress, "iommu_group"))
```

Comment:
Do we have similar functionality in `go-nvlib`?

Reply:
So we have it tracked in `NvidiaPciDevice`  here: https://github.com/NVIDIA/go-nvlib/blob/main/pkg/nvpci/nvpci.go. I'm guessing `nvpci` is the sysfs way of discovering GPUs. But `nvlib` doesn't have it.

Btw one of the implicit assumptions I havent captured here is that `iommu` needs to be enabled in the kernel for us to be able to do gpu-passthrough.  In that case we would hit the err in L171 and return empty iommu group.